server/src/test/java/org/elasticsearch/transport/TcpTransportTests.java

```diff
@@ -354,4 +366,101 @@ public void testHTTPResponse() throws IOException {
+        testExceptionHandling(new StreamCorruptedException("simulated"),
+            new MockLogAppender.SeenEventExpectation("message", "org.elasticsearch.transport.TcpTransport",
+                Level.WARN, "simulated, [*], closing connection"));
+    }
+
+    private void testExceptionHandling(Exception exception,
+                                       MockLogAppender.LoggingExpectation... expectations) throws IllegalAccessException, IOException {
+        testExceptionHandling(true, exception, true, expectations);
+    }
+
+    private void testExceptionHandling(boolean startTransport, Exception exception, boolean expectClosed,
+                                       MockLogAppender.LoggingExpectation... expectations) throws IllegalAccessException, IOException {
+        final TestThreadPool testThreadPool = new TestThreadPool("test");
+        TcpTransport tcpTransport = null;
+        MockLogAppender appender = new MockLogAppender();
+
+        try {
+            appender.start();
+
+            Loggers.addAppender(LogManager.getLogger(TcpTransport.class), appender);
+            for (MockLogAppender.LoggingExpectation expectation : expectations) {
+                appender.addExpectation(expectation);
+            }
+
+            tcpTransport = new TcpTransport(Settings.EMPTY, Version.CURRENT, testThreadPool,
+                new MockPageCacheRecycler(Settings.EMPTY),
+                new NoneCircuitBreakerService(), writableRegistry(), new NetworkService(Collections.emptyList())) {
+
+                @Override
+                protected TcpServerChannel bind(String name, InetSocketAddress address) {
+                    throw new UnsupportedOperationException();
+                }
+
+                @Override
+                protected TcpChannel initiateChannel(DiscoveryNode node) {
+                    throw new UnsupportedOperationException();
+                }
+
+                @Override
+                protected void stopInternal() {
+                }
+            };
+
+            if (startTransport) {
+                tcpTransport.start();
+            }
+
+            final FakeTcpChannel channel = new FakeTcpChannel();
+            final PlainActionFuture<Void> listener = new PlainActionFuture<>();
+            channel.addCloseListener(listener);
+
+            tcpTransport.onException(channel, exception);
+
+            if (expectClosed) {
+                assertTrue(listener.isDone());
+                assertThat(listener.actionGet(), nullValue());
+            } else {
+                assertFalse(listener.isDone());
+            }
+
+            appender.assertAllExpectationsMatched();
+
+        } finally {
+            Loggers.removeAppender(LogManager.getLogger(TcpTransport.class), appender);
+            appender.stop();
+            IOUtils.close(tcpTransport);
+            testThreadPool.shutdown();
```

Comment:
NIT: Maybe use `org.elasticsearch.threadpool.ThreadPool#terminate(java.util.concurrent.ExecutorService, long, java.util.concurrent.TimeUnit)` here so we don't get the leaked thread warning when running this one?

Reply:
TIL. Yes, done in f0e813b23be.